src/dbnode/network/server/tchannelthrift/node/service.go

```diff
@@ -729,153 +728,291 @@ func (s *service) FetchTagged(tctx thrift.Context, req *rpc.FetchTaggedRequest)
+func (i *fetchTaggedResultsIter) Exhaustive() bool {
+	return i.exhaustive
+}
+
+func (i *fetchTaggedResultsIter) Namespace() ident.ID {
+	return i.nsID
+}
 
-		enc := s.pools.tagEncoder.Get()
-		ctx.RegisterFinalizer(enc)
-		encodedTags, err := s.encodeTags(enc, tags)
-		if err != nil { // This is an invariant, should never happen
-			return tterrors.NewInternalError(err)
+func (i *fetchTaggedResultsIter) Next(ctx context.Context) bool {
+	if i.idx >= len(i.queryResults) {
+		return false
+	}
+	// TODO(rhall): don't request all series blocks at once.
+	if i.idx == 0 {
+		for _, entry := range i.queryResults { // nolint: gocritic
+			result := IDResult{
+				queryResult: entry,
+				docReader:   i.docReader,
+				tagEncoder:  i.tagEncoder,
+				iOpts:       i.iOpts,
+			}
+			if i.fetchData {
+				// NB(r): Use a bytes ID here so that this ID doesn't need to be
+				// copied by the blockRetriever in the streamRequest method when
+				// it checks if the ID is finalizeable or not with IsNoFinalize.
+				id := ident.BytesID(result.queryResult.Key())
+				result.blockReaders, i.err = i.db.ReadEncoded(ctx, i.nsID, id, i.startInclusive, i.endExclusive)
+				if i.err != nil {
+					return true
+				}
+			}
+			i.idResults = append(i.idResults, result)
 		}
+	}
+	i.cur = i.idResults[i.idx]
+	i.idx++
+	return true
+}
 
-		elem := &rpc.FetchTaggedIDResult_{
-			NameSpace:   nsIDBytes,
-			ID:          id.Bytes(),
-			EncodedTags: encodedTags.Bytes(),
-		}
-		response.Elements = append(response.Elements, elem)
-		if !fetchData {
-			continue
-		}
+func (i *fetchTaggedResultsIter) Err() error {
+	return i.err
+}
 
-		encoded, err := db.ReadEncoded(ctx, nsID, id,
-			opts.StartInclusive, opts.EndExclusive)
-		if err != nil {
-			return convert.ToRPCError(err)
-		} else {
-			encodedDataResults[idx] = encoded
-		}
+func (i *fetchTaggedResultsIter) Current() IDResult {
+	return i.cur
+}
+
+// IDResult is the FetchTagged result for a series ID.
+type IDResult struct {
+	queryResult  index.ResultsMapEntry
+	docReader    *docs.EncodedDocumentReader
+	tagEncoder   serialize.TagEncoder
+	blockReaders [][]xio.BlockReader
+	iOpts        instrument.Options
+}
+
+// ID returns the series ID.
+func (i *IDResult) ID() []byte {
+	return i.queryResult.Key()
+}
+
+// WriteTags writes the encoded tags to provided slice. Callers must use the returned reference in case the slice needed
+// to grow, just like append().
+func (i *IDResult) WriteTags(result []byte) ([]byte, error) {
+	metadata, err := docs.MetadataFromDocument(i.queryResult.Value(), i.docReader)
+	if err != nil {
+		return nil, err
 	}
-	return nil
+	tags := idxconvert.ToSeriesTags(metadata, idxconvert.Opts{NoClone: true})
+	encodedTags, err := encodeTags(i.tagEncoder, tags, i.iOpts)
+	if err != nil { // This is an invariant, should never happen
+		return nil, tterrors.NewInternalError(err)
+	}
+	result = append(result, encodedTags.Bytes()...)
```

Comment:
nit: A slightly more defensive programming approach here is to slice the `result` into `result[:0]` so that if caller accidently passed a full buffer for reuse that hadn't been resized it will overwrite with the length reset but still able to use the allocated capacity. Also very nitty, we usually the name `dst` for a byte slice that is to be written into.

e.g.
```go
func (i *IDResult) WriteTags(result []byte) ([]byte, error) {
  // .... other
  dst = append(dst[:0], encodedTags.Bytes()...)
```